### DISCUSSION TITLE
Implementing Librashader in a Rust emulator

### DISCUSSION THREAD

**Discussion by dbalsom** [SnowflakePowered/librashader#22] [Q&A] [answered]
I have a PC emulator written in Rust, using winit, wgpu and egui.
https://github.com/dbalsom/martypc

I'd love to implement Librashader, but I confess I am a bit confused as to how I'd go about doing it.  There are examples under each 'runtime' but I don't see an example of just importing the librashader crate and going from there. 

The docs say 

> The basic workflow involves parsing a ShaderPreset, which can then be used to construct a FilterChain. All shaders will then be compiled, after which FilterChain::frame can be called with appropriate input and output parameters to draw a frame with the shader effect applied.

How do I get a FilterChain? It seems that it is defined for each runtime. Do I have to choose a runtime statically? Is there a way to query what runtimes are supported on a specific platform?

**Comment by chyyran** [comment]
WGPU might be a little more complicated for the fact that the current runtimes work with the underlying surface of a HAL but it's [probably doable with some arm twisting.](https://wgpu.rs/doc/wgpu_core/hal_api/trait.HalApi.html). You'll have to figure out which graphics backend is being used at runtime, and select for the appropriate librashader `FilterChain` struct.


Getting the `FilterChain` itself differs depending on the runtime. Here's how it can work for Vulkan, assuming you're able to get the underlying `vkPhysicalDevice`, `vkInstance` and `vkDevice` handles.

```rust
use librashader::runtime::vk::*;

fn load_vulkan_filter_chain(path: impl AsRef<Path>, pd: ash::vk::PhysicalDevice, instance: ash::Instance, device: ash::Device) -> Result<FilterChain, FilterChainError> {
  let preset = ShaderPreset::try_parse(path)?;
  FilterChain::load_from_preset(preset, (pd, instance, device), None)
}
```

Then after rendering, but before you present, call [`FilterChain::frame`](https://docs.rs/librashader/latest/librashader/runtime/vk/struct.FilterChain.html#method.frame).

The rough structure is the same for every runtime, 

Usually in every runtime you'll find a working minimal example in the `test` folder. Here's the one for [the Vulkan runtime](https://github.com/SnowflakePowered/librashader/tree/master/librashader-runtime-vk/tests).

**Comment by chyyran** [comment]
I'm starting a branch over at https://github.com/SnowflakePowered/librashader/tree/feat-wgpu-runtime to build a WGPU runtime, but the only (kind of big) issue is that it needs `wgpu::Features::SPIRV_SHADER_PASSTHROUGH`. which pretty much makes the only backend compatible with WGPU the Vulkan one. Given that I'm not sure if it's worth the work to maintain such a backend especially if the low-level Vulkan backend is already available.

This is pretty much because of https://github.com/gfx-rs/wgpu/issues/4342, since most if not all RetroArch shaders rely on combined image samplers.

**Comment by dbalsom** [comment]
Thanks for the info.

I'm working on a display abstraction for my emulator so I'm not necessarily committed to rendering with wgpu for a given window.  I can create a new window and attach anything that I've implemented my RenderBackend trait.

Here's what I'm thinking:
I have a main window, which I will initialize with winit/wgpu and render the primary video card and egui to, as normal. 
I will have a menu/config option to create a separate window (or windows) sourced from the specified video card(s), and attach a RenderBackend implemented for librashader.

Before creating this backend I can query the main wgpu instance to see what backend it selects and then attempt to create a librashader FilterChain of the same type. But this assumes I can render to an existing Winit window handle, or retrieve the raw window handle from the new window and use to render. I'd rather avoid having a window created externally as it is convenient for all my windows to share the same event loop.

A wgpu backend would probably be pretty convenient at least to me, even if it only worked on Vulkan. My emulator's wgpu frontend already pretty much performs poorly on any backend that isn't Vulkan already. I have a built in CRT shader, so I assume anyone wanting to play with Retroarch shader presets probably has a decent graphics card that will support Vulkan.  It would especially be nice if it would cut down on the number of additional dependencies required.

I'll play around with the Vulkan runtime example and see if I can get this working.   

Thanks for creating this, by the way - shader support has been one of the more popular feature requests I've received and I had no idea how I was going to implement that on my own.

**Comment by chyyran** [comment]
Just to give a quick update, it was apparent fairly early on that simply loading SPIR-V shaders wasn’t going to work for WGPU, since it has no API to actually bind combined image samplers. 

To that end, I’m working on a couple of SPIR-V transforms over at https://github.com/chyyran/kirisaku  that will split up combined image samplers and put the samplers in bind group 1. I’ll also have to put push constants into their own UBO since WGSL doesn’t support those either and it’s a much more trivial transform than splitting up `sampler2D`s.

That should mean though that the WGPU runtime will hopefully be compatible with more than just Vulkan, and also WebAssembly-compatible since it’ll be working with valid WGSL. May need to tweak which features get compiled in for WASM targets though.

**Comment by chyyran** [comment, marked as answer]
As of https://github.com/SnowflakePowered/librashader/pull/24 librashader now provides a [wgpu runtime](https://docs.rs/librashader/0.2.0-beta.6/librashader/runtime/wgpu/index.html).
